litmus-portal/graphql-server/pkg/database/mongodb/operations.go

```diff
@@ -44,32 +44,34 @@ func UpdateCluster(query bson.D, update bson.D) error {
 	return nil
 }
 
-func UpsertWorkflowRun(workflow_id string, wfRun WorkflowRun) error {
+func UpsertWorkflowRun(workflow_id string, wfRun WorkflowRun) (int, error) {
 	opts := options.Update().SetUpsert(true)
 	ctx, _ := context.WithTimeout(backgroundContext, 10*time.Second)
 
 	count, err := workflowCollection.CountDocuments(ctx, bson.M{"workflow_id": workflow_id, "workflow_runs.workflow_run_id": wfRun.WorkflowRunID})
 	if err != nil {
-		return err
+		return 0, err
 	}
 
+	updateCount := 1
 	if count == 0 {
 		filter := bson.M{"workflow_id": workflow_id}
 		update := bson.M{"$push": bson.M{"workflow_runs": wfRun}}
 		_, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
 		if err != nil {
```

Comment:
```suggestion
	updateCount := 0
	if count == 0 {
		filter := bson.M{"workflow_id": workflow_id}
		update := bson.M{"$push": bson.M{"workflow_runs": wfRun}}
		updateResp, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
		if err != nil {
		    return 0, err
		}
		updateCount = int(updateResp.MatchedCount)
```

Reply:
This is not needed, the updateCount is required for only already existing workflow_runs

Reply:
In case of `count == 0`, it is also updated. It is required if updates fail during that time

Reply:
If the update fails we will return the error. If there is no error then that means the update succeeded the updateCount is irrelevant in that case. For count==0 case this patch is irrelevant altogether.

Reply:
If updateCount is specific for existing workflow_runs make it boolean. It should operate based on true/false.

Reply:
```
		if int(updateResp.MatchedCount) > 0 {
                         isExisting = true
                 }

```

Reply:
No, the logic is not about whether the workflow_run exists but whether a record is updated. If there is an update the updateCount will always be 1. If there is no update it will be 0.  That is why in the error paths I have mentioned updateCount as 0 because no record is being updated, besides returning the error. The default value is 1 because for count==0 the updateCount will always be 1 (or 0 in error) as only 1 workflow_run can been updated/inserted. I think keeping it integer suits the usecase and logic.

Reply:
> If the update fails we will return the error. If there is no error then that means the update succeeded the updateCount is irrelevant in that case. For count==0 case this patch is irrelevant altogether.

In some cases, MongoDB will not return err, if update query params didn't match(It will return matchcount == 0). In that case update count remains 1

Reply:
In count == 0, the updateCount will not always return 0 in err, as mention the case here- https://github.com/litmuschaos/litmus/pull/2312/files#r516842184

Reply:
That case wasn't handled previously? If this is a possibility it should've been handled. Though I doubt this is possible because the code always makes a db entry for the workflow first and only after that succeeds it sends the workflow to the agent. The wf_id if provided by the agent should always match.

Reply:
If that's a possibility we should be throwing an error when the wf_id doesn't match.

Reply:
It wasn't handled previously, It is not about wf_id, if workflos_runs  is not available it will return `Modified: 0`

Reply:
Did not get you, the query is about workflow_id right, so the update is dependant on the workflow_id itself. The workflow_runs array can be empty but that shouldn't cause any problems as we are pushing a new workflow inside it. The update count should be 1 as long as the wf_id matches

Reply:
The thing is don't initialize updateCount == 1 at first, we can't predict the mongodb response.
Set to 1 once you got the response from the mongodb

Reply:
Okay I'll update the code to throw an error if the matchCount is 0 for the wf_id. That should solve both the problems

Reply:
Updated, it should now throw error if not workflow_id match is found.